frame/fee-market/src/lib.rs

```diff
@@ -296,122 +312,109 @@ pub mod pallet {
 	/// An important update in this pallet, need to update market information in the following cases:
 	///
-	/// - When new relayer enroll
-	/// - When enrolled relayer wants to update relaying fee
-	/// - When enrolled relayer wants to cancel enrollment
-	/// - When some enrolled relayer's collateral below MiniumLockCollateral, might trigger market update
-	pub fn update_market() {
-		// Sort all enrolled relayers firstly.
+	/// - New relayer enroll.
+	/// - The enrolled relayer wants to update fee or order capacity.
+	/// - The enrolled relayer wants to cancel enrollment.
+	/// - The order didn't confirm in-time, slash occurred.
+	pub(crate) fn update_market() {
+		// Sort all enrolled relayers who are able to accept orders.
 		let mut relayers: Vec<Relayer<T::AccountId, RingBalance<T>>> = <Relayers<T>>::get()
 			.iter()
 			.map(RelayersMap::<T>::get)
+			.filter(|r| Self::usable_order_capacity(&r.id) >= 1)
 			.collect();
-		relayers.sort();
 
 		// Select the first `AssignedRelayersNumber` relayers as AssignedRelayer.
-		// Only when total relayer's number greater than `AssignedRelayersNumber`, selection happens.
 		let assigned_relayers_len = T::AssignedRelayersNumber::get() as usize;
 		if relayers.len() >= assigned_relayers_len {
-			let mut assigned_relayers = Vec::with_capacity(assigned_relayers_len);
-			for i in 0..assigned_relayers_len {
-				if let Some(r) = relayers.get(i) {
-					assigned_relayers.push(r);
-				}
-			}
-			<AssignedRelayersStorage<T>>::put(assigned_relayers);
+			relayers.sort();
+
+			let assigned_relayers: Vec<_> = relayers.iter().take(assigned_relayers_len).collect();
+			<AssignedRelayers<T>>::put(assigned_relayers);
+		} else {
+			// The market fee comes from the last item in AssignedRelayers,
+			// It's would be essential to wipe this storage if relayers not enough.
+			<AssignedRelayers<T>>::kill();
 		}
 	}
 
-	/// Update relayer locked collateral, then update market fee, this will changes RelayersMap storage.
-	pub fn update_collateral(who: &T::AccountId, new_collateral: RingBalance<T>) {
-		// Ensure always at least `AssignedRelayersNumber` relayers are able to provide market fee.
-		if new_collateral < T::MiniumLockCollateral::get()
-			&& <Relayers<T>>::get().len() > T::AssignedRelayersNumber::get() as usize
-		{
-			Self::remove_enrolled_relayer(who);
-			return;
-		}
-
-		let _ = T::RingCurrency::extend_lock(
+	/// Update relayer after slash occurred, this will changes RelayersMap storage. (Update market needed)
+	pub(crate) fn update_relayer_after_slash(who: &T::AccountId, new_collateral: RingBalance<T>) {
+		T::RingCurrency::set_lock(
 			T::LockId::get(),
-			who,
-			new_collateral,
+			&who,
+			LockFor::Common {
+				amount: new_collateral,
+			},
 			WithdrawReasons::all(),
 		);
 		<RelayersMap<T>>::mutate(who.clone(), |relayer| {
 			relayer.collateral = new_collateral;
 		});
+
+		if Self::usable_order_capacity(&who) == 0 {
+			Self::remove_enrolled_relayer(who);
+			return;
+		}
```

Comment:
Do not remove relayer here, let's relayer remove by themselves.

And this relay might have other order in working.

Reply:
Done.